source/bigint.c

```diff
@@ -448,254 +456,287 @@ int aws_bigint_bytebuf_debug_output(const struct aws_bigint *bigint, struct aws_
+    size_t buffer_offset = byte_length - bytes_to_write;
+    uint32_t *raw_digits = bigint->digits.data;
+    for (size_t i = 0; i < bytes_to_write; ++i) {
+        size_t digit_index = i / BYTES_PER_BASE_DIGIT;
+        uint32_t digit = raw_digits[digit_index];
+        uint32_t digit_byte_index = i % BYTES_PER_BASE_DIGIT;
+        uint8_t digit_byte = (digit >> (BITS_PER_BYTE * digit_byte_index)) & 0xFF;
+        size_t buffer_index = bytes_to_write - i - 1 + buffer_offset;
 
-            if (should_write) {
-                if (aws_byte_buf_append(buffer, &digit_bytes_cursor)) {
-                    goto done;
-                }
-            }
-        }
+        raw_buffer[buffer_index] = digit_byte;
     }
 
     result = AWS_OP_SUCCESS;
 
 done:
 
-    AWS_POSTCONDITION(aws_bigint_is_valid(bigint));
-    AWS_POSTCONDITION(aws_byte_buf_is_valid(buffer));
+    AWS_FATAL_POSTCONDITION(s_aws_bigint_is_valid(bigint));
+    AWS_FATAL_POSTCONDITION(aws_byte_buf_is_valid(buffer));
 
     return result;
 }
 
-#if _MSC_VER
-#    pragma warning(pop)
-#endif
-
 bool aws_bigint_is_negative(const struct aws_bigint *bigint) {
-    AWS_PRECONDITION(aws_bigint_is_valid(bigint));
+    AWS_FATAL_PRECONDITION(s_aws_bigint_is_valid(bigint));
 
     return bigint->sign < 0;
 }
 
+static int8_t s_aws_bigint_is_zero(const struct aws_bigint *bigint) {
+    AWS_FATAL_PRECONDITION(s_aws_bigint_is_valid(bigint));
+
+    volatile int8_t is_zero = 1;
+    size_t digit_count = aws_array_list_length(&bigint->digits);
+    uint32_t *raw_digits = bigint->digits.data;
+
+    for (size_t i = 0; i < digit_count; ++i) {
+        int64_t digit = (int64_t)raw_digits[i];
+        is_zero &= ((digit - 1) >> 63) & 0x01;
+    }
+
+    return is_zero;
+}
+
 bool aws_bigint_is_positive(const struct aws_bigint *bigint) {
-    AWS_PRECONDITION(aws_bigint_is_valid(bigint));
+    AWS_FATAL_PRECONDITION(s_aws_bigint_is_valid(bigint));
+
+    volatile uint8_t sign_one = 1 - ((bigint->sign >> 1) & 0x01);
 
-    return bigint->sign > 0 && !aws_bigint_is_zero(bigint);
+    return sign_one & (1 - s_aws_bigint_is_zero(bigint));
 }
 
 bool aws_bigint_is_zero(const struct aws_bigint *bigint) {
-    AWS_PRECONDITION(aws_bigint_is_valid(bigint));
+    AWS_FATAL_PRECONDITION(s_aws_bigint_is_valid(bigint));
 
-    if (bigint->sign < 0) {
-        return false;
-    }
+    return s_aws_bigint_is_zero(bigint);
+}
 
-    size_t digit_count = aws_array_list_length(&bigint->digits);
-    if (digit_count != 1) {
-        return false;
-    }
+void aws_bigint_negate(struct aws_bigint *bigint) {
+    AWS_FATAL_PRECONDITION(s_aws_bigint_is_valid(bigint));
+
+    volatile int8_t is_zero = s_aws_bigint_is_zero(bigint);
 
-    uint32_t digit = 0;
-    aws_array_list_get_at(&bigint->digits, &digit, 0);
+    /* 1 if the the number is 0; -1 if the number is non-zero */
+    volatile int8_t negation_factor = is_zero + is_zero - 1;
+    bigint->sign *= negation_factor; /* is multiplication kosher? */
```

Comment:
Why wouldn't it be kosher? It's constant time too for [-1|1]

Reply:
Probably fine, found some stuff hinting that even 32, 32 -> 64 was not constant time on some archivectures (early arm).

Reply:
Dropped, if constant-time multiplication ever becomes an issue will think about then I guess.